File: langs/ruby/lib/dotlyte/typed_config.rb
# frozen_string_literal: true

require "set"

module Dotlyte
  # Typed configuration builder with schema-driven validation and type coercion.
  #
  # A FieldDescriptor is a Hash with keys:
  #   :type       - "string", "integer", "float", "boolean", "array" (default: "string")
  #   :required   - Boolean (default: false)
  #   :default    - Default value if not set
  #   :enum       - Array of allowed values
  #   :min        - Minimum value (numeric types)
  #   :max        - Maximum value (numeric types)
  #   :sensitive  - Boolean, marks the field as sensitive (default: false)
  #   :doc        - Documentation string
  #
  # @example
  #   config = Dotlyte::TypedConfig.create(
  #     "PORT"     => { type: "integer", required: true, min: 1, max: 65535 },
  #     "DEBUG"    => { type: "boolean", default: false },
  #     "LOG_LEVEL" => { type: "string", enum: %w[debug info warn error], default: "info" }
  #   )
  #   config["PORT"]  #=> 8080
  class TypedConfig
    BOOLEAN_TRUE  = Set.new(%w[true yes 1 on]).freeze
    BOOLEAN_FALSE = Set.new(%w[false no 0 off]).freeze

    # Create a typed configuration from environment variables.
    #
    # @param schema [Hash{String => Hash}] key => FieldDescriptor mapping
    # @param skip_validation [Boolean] skip validation after coercion (default: false)
    # @param on_secret_access [Proc, nil] callback invoked when a sensitive key is accessed
    # @return [Hash] frozen Hash-like object with coerced, validated values
    # @raise [Dotlyte::Error] if validation fails
    def self.create(schema, skip_validation: false, on_secret_access: nil)
      new(schema, skip_validation: skip_validation, on_secret_access: on_secret_access).build
    end

    private_class_method :new

    # @api private
    def initialize(schema, skip_validation:, on_secret_access:)
      @schema = schema
      @skip_validation = skip_validation
      @on_secret_access = on_secret_access
    end

    # @api private
    # @return [Hash] frozen configuration hash
    def build
      errors = []
      result = {}

      @schema.each do |key, descriptor|
        descriptor = normalize_descriptor(descriptor)
        raw = ENV[key]

        # Apply default if not present
        if raw.nil? || raw.empty?
          if descriptor[:default] != nil # rubocop:disable Style/NonNilCheck
            result[key] = descriptor[:default]
            next
          elsif descriptor[:required]
            errors << "Required config key '#{key}' is missing. Set it as an environment variable."
            next
          else
            next
          end
        end

        # Coerce type
        begin
          value = coerce_value(raw, descriptor[:type], key)
        rescue Error => e
          errors << e.message
          next
        end

        # Validate enum
        if descriptor[:enum] && !descriptor[:enum].include?(value)
          errors << "Config key '#{key}': value #{value.inspect} is not in allowed values: #{descriptor[:enum].inspect}"
          next
        end

        # Validate min/max
        if value.is_a?(Numeric)
          if descriptor[:min] && value < descriptor[:min]
            errors << "Config key '#{key}': value #{value} is less than minimum #{descriptor[:min]}"
          end
          if descriptor[:max] && value > descriptor[:max]
            errors << "Config key '#{key}': value #{value} is greater than maximum #{descriptor[:max]}"
          end
        end

        result[key] = value
      end

      unless @skip_validation || errors.empty?
        raise Error, "TypedConfig validation failed:\n  - #{errors.join("\n  - ")}"
      end

      # Wrap with sensitive-access tracking if needed
      if @on_secret_access
        sensitive_keys = @schema.select { |_, d| d[:sensitive] }.keys.to_set
        build_sensitive_proxy(result, sensitive_keys)
      else
        result.freeze
      end
    end

    private

    # Normalize a FieldDescriptor hash with defaults.
    #
    # @param desc [Hash] raw descriptor
    # @return [Hash] normalized descriptor
    def normalize_descriptor(desc)
      {
        type: desc[:type] || "string",
        required: desc.fetch(:required, false),
        default: desc.key?(:default) ? desc[:default] : nil,
        enum: desc[:enum],
        min: desc[:min],
        max: desc[:max],
        sensitive: desc.fetch(:sensitive, false),
        doc: desc[:doc]
      }
    end

    # Coerce a string value to the expected type.
    #
    # @param raw [String] raw string value from ENV
    # @param type [String] expected type name
    # @param key [String] config key (for error messages)
    # @return [Object] coerced value
    # @raise [Dotlyte::Error] if coercion fails
    def coerce_value(raw, type, key)
      case type
      when "string"
        raw
      when "integer"
        unless raw.match?(/\A-?\d+\z/)
          raise Error, "Config key '#{key}': cannot convert #{raw.inspect} to integer"
        end
        Integer(raw)
      when "float"
        unless raw.match?(/\A-?\d+(\.\d+)?\z/)
          raise Error, "Config key '#{key}': cannot convert #{raw.inspect} to float"
        end
        Float(raw)
      when "boolean"
        lower = raw.downcase
        return true if BOOLEAN_TRUE.include?(lower)
        return false if BOOLEAN_FALSE.include?(lower)

        raise Error, "Config key '#{key}': cannot convert #{raw.inspect} to boolean"
      when "array"
        raw.split(",").map(&:strip)
      else
        raise Error, "Config key '#{key}': unknown type '#{type}'"
      end
    end

    # Build a proxy Hash that calls on_secret_access for sensitive keys.
    #
    # @param data [Hash] the config data
    # @param sensitive_keys [Set<String>] keys marked as sensitive
    # @return [Object] frozen proxy object
    def build_sensitive_proxy(data, sensitive_keys)
      callback = @on_secret_access
      frozen_data = data.freeze

      proxy = Object.new

      # Define [] accessor
      proxy.define_singleton_method(:[]) do |key|
        callback.call(key) if sensitive_keys.include?(key)
        frozen_data[key]
      end

      # Define key? method
      proxy.define_singleton_method(:key?) do |key|
        frozen_data.key?(key)
      end

      # Define keys method
      proxy.define_singleton_method(:keys) do
        frozen_data.keys
      end

      # Define to_h
      proxy.define_singleton_method(:to_h) do
        frozen_data.dup
      end

      # Define fetch
      proxy.define_singleton_method(:fetch) do |key, *args|
        callback.call(key) if sensitive_keys.include?(key)
        frozen_data.fetch(key, *args)
      end

      proxy.freeze
    end
  end
end

File: langs/dotnet/src/Dotlyte/TypedConfig.cs
using System.Globalization;

namespace Dotlyte;

/// <summary>
/// Describes a single configuration field's type, constraints, and metadata.
/// </summary>
public sealed class FieldDescriptor
{
    /// <summary>Expected type: "string", "integer", "boolean", "number", "array".</summary>
    public string Type { get; init; } = "string";

    /// <summary>Whether the field is required.</summary>
    public bool Required { get; init; }

    /// <summary>Default value applied when the field is missing.</summary>
    public object? Default { get; init; }

    /// <summary>Allowed values for the field.</summary>
    public string[]? Enum { get; init; }

    /// <summary>Minimum numeric value.</summary>
    public double? Min { get; init; }

    /// <summary>Maximum numeric value.</summary>
    public double? Max { get; init; }

    /// <summary>Whether the field contains sensitive data.</summary>
    public bool Sensitive { get; init; }

    /// <summary>Documentation string for the field.</summary>
    public string? Doc { get; init; }
}

/// <summary>
/// Creates a strongly-typed configuration dictionary from a schema of <see cref="FieldDescriptor"/>
/// entries, reading values from environment variables with automatic type coercion and validation.
/// </summary>
public static class TypedConfig
{
    private static readonly HashSet<string> TrueValues = new(StringComparer.OrdinalIgnoreCase)
        { "true", "yes", "1", "on" };

    private static readonly HashSet<string> FalseValues = new(StringComparer.OrdinalIgnoreCase)
        { "false", "no", "0", "off" };

    private static readonly HashSet<string> NullValues = new(StringComparer.OrdinalIgnoreCase)
        { "null", "none", "nil", "" };

    /// <summary>
    /// Create a typed configuration dictionary from a schema of field descriptors.
    /// Values are read from <see cref="Environment.GetEnvironmentVariable(string)"/>
    /// and coerced to the declared type.
    /// </summary>
    /// <param name="schema">Map of key names to their field descriptors.</param>
    /// <param name="skipValidation">When <c>true</c>, skip enum/min/max validation.</param>
    /// <param name="onSecretAccess">Optional callback invoked when a sensitive field is accessed.</param>
    /// <returns>A dictionary of coerced, validated configuration values.</returns>
    /// <exception cref="DotlyteException">Thrown when a required key is missing or validation fails.</exception>
    public static Dictionary<string, object?> Create(
        Dictionary<string, FieldDescriptor> schema,
        bool skipValidation = false,
        Action<string>? onSecretAccess = null)
    {
        var result = new Dictionary<string, object?>();

        foreach (var (key, descriptor) in schema)
        {
            var envKey = key.Replace(".", "_").Replace("-", "_").ToUpperInvariant();
            var raw = Environment.GetEnvironmentVariable(envKey);

            object? value;
            if (raw is not null)
            {
                value = CoerceToType(raw, descriptor.Type, key);
            }
            else if (descriptor.Default is not null)
            {
                value = descriptor.Default;
            }
            else
            {
                value = null;
            }

            // Required check
            if (descriptor.Required && value is null)
            {
                throw new DotlyteException(
                    $"Required configuration key '{key}' is missing. " +
                    $"Set environment variable '{envKey}' or provide a default.", key);
            }

            // Validation (unless skipped)
            if (!skipValidation && value is not null)
            {
                ValidateField(key, value, descriptor);
            }

            // Notify on sensitive access
            if (descriptor.Sensitive && onSecretAccess is not null && value is not null)
            {
                onSecretAccess(key);
            }

            result[key] = value;
        }

        return result;
    }

    private static object? CoerceToType(string raw, string type, string key)
    {
        var trimmed = raw.Trim();

        if (NullValues.Contains(trimmed))
            return null;

        return type.ToLowerInvariant() switch
        {
            "string" => trimmed,
            "boolean" or "bool" => CoerceBoolean(trimmed, key),
            "integer" or "int" => CoerceInteger(trimmed, key),
            "number" or "float" or "double" => CoerceNumber(trimmed, key),
            "array" or "list" => CoerceArray(trimmed),
            _ => Coercion.Coerce(trimmed),
        };
    }

    private static bool CoerceBoolean(string value, string key)
    {
        if (TrueValues.Contains(value)) return true;
        if (FalseValues.Contains(value)) return false;
        throw new DotlyteException(
            $"Cannot coerce '{value}' to boolean for key '{key}'", key);
    }

    private static long CoerceInteger(string value, string key)
    {
        if (long.TryParse(value, NumberStyles.Integer, CultureInfo.InvariantCulture, out var result))
            return result;
        throw new DotlyteException(
            $"Cannot coerce '{value}' to integer for key '{key}'", key);
    }

    private static double CoerceNumber(string value, string key)
    {
        if (double.TryParse(value, NumberStyles.Float | NumberStyles.AllowThousands,
            CultureInfo.InvariantCulture, out var result))
            return result;
        throw new DotlyteException(
            $"Cannot coerce '{value}' to number for key '{key}'", key);
    }

    private static List<object?> CoerceArray(string value)
    {
        return value.Split(',')
            .Select(item => Coercion.Coerce(item.Trim()))
            .ToList();
    }

    private static void ValidateField(string key, object value, FieldDescriptor descriptor)
    {
        // Enum validation
        if (descriptor.Enum is not null)
        {
            var strValue = value.ToString();
            if (!descriptor.Enum.Contains(strValue))
            {
                throw new DotlyteException(
                    $"Value '{strValue}' for key '{key}' is not in allowed values: [{string.Join(", ", descriptor.Enum)}]",
                    key);
            }
        }

        // Min/Max validation
        if (descriptor.Min is not null || descriptor.Max is not null)
        {
            if (TryGetNumeric(value, out var num))
            {
                if (descriptor.Min is not null && num < descriptor.Min.Value)
                {
                    throw new DotlyteException(
                        $"Value {num} for key '{key}' is below minimum {descriptor.Min.Value}",
                        key);
                }

                if (descriptor.Max is not null && num > descriptor.Max.Value)
                {
                    throw new DotlyteException(
                        $"Value {num} for key '{key}' exceeds maximum {descriptor.Max.Value}",
                        key);
                }
            }
        }
    }

    private static bool TryGetNumeric(object value, out double result)
    {
        result = 0;
        try
        {
            result = Convert.ToDouble(value, CultureInfo.InvariantCulture);
            return true;
        }
        catch
        {
            return false;
        }
    }
}
